src/aolayer.cpp

```diff
@@ -269,14 +269,17 @@ void CharLayer::start_playback(QString p_image)
 
 void AOLayer::start_playback(QString p_image)
 {
+  QMutexLocker locker(&mutex);
+  if (frame_loader.isRunning())
+    exit_loop = true; // tell the loader to stop, we have a new image to load
```

Comment:
just realized how a use after free can happen with your old code - not all animations run to completion. setting a stop variable won't necessarily stop a use after free, you need to wait for the thread to get out of the loop completely

Reply:
QtConcurrent::run doesn't have a way of canceling the thread, so this is the workaround for that. The use after free was a problem with an earlier version of this code, I've never had it happen to this iteration.

Reply:
never mind -- you've used a QMutexLocker which locks on RAII semantics. this will work